File: app/src/main/java/com/example/chuckjokes/Jokes/JokesFragment.java
package com.example.chuckjokes.Jokes;

import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.example.chuckjokes.Jokes.presenter.JokesPresenter;
import com.example.chuckjokes.Jokes.presenter.RecyclerAdapter;
import com.example.chuckjokes.MainActivity;

import java.lang.ref.WeakReference;
import com.example.chuckjokes.databinding.JokesFragmentBinding;

import org.jetbrains.annotations.NotNull;

public class JokesFragment extends Fragment {

    private final WeakReference<MainActivity> mainActivityWeakReference;
    private JokesFragmentBinding binding;
    private final JokesPresenter presenter;

    public JokesFragment(MainActivity mainActivity) {
        this.mainActivityWeakReference = new WeakReference<>(mainActivity);
        this.presenter = new JokesPresenter(this);
    }

    @Override
    public View onCreateView(@NotNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        binding = JokesFragmentBinding.inflate(inflater);
        presenter.init();
        return binding.JokesList;
    }

    public void setLayoutManager(LinearLayoutManager linearLayoutManager) {
        binding.JokesList.setLayoutManager(linearLayoutManager);
    }

    public void setAdapter(RecyclerAdapter recyclerAdapter) {
        binding.JokesList.setAdapter(recyclerAdapter);
    }

    public Context getRecyclerContext() {
        return binding.JokesList.getContext();
    }

    public void addItemDecoration(DividerItemDecoration dividerItemDecoration) {
        binding.JokesList.addItemDecoration(dividerItemDecoration);
    }

    public void addOnScrollListener(RecyclerView.OnScrollListener onScrollListener) {
        binding.JokesList.addOnScrollListener(onScrollListener);
    }

    public WeakReference<MainActivity> getMainActivityWeakReference() {
        return mainActivityWeakReference;
    }
}

File: app/src/main/java/com/example/chuckjokes/jokes/JokesFragment.kt
package com.example.chuckjokes.jokes

import android.view.LayoutInflater
import android.view.ViewGroup
import android.os.Bundle
import android.view.View
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.core.content.res.ResourcesCompat
import androidx.fragment.app.Fragment
import com.example.chuckjokes.R
import androidx.recyclerview.widget.RecyclerView
import com.example.chuckjokes.Shared
import com.example.chuckjokes.databinding.JokesFragmentBinding
import com.example.chuckjokes.main.MainActivity
import com.example.chuckjokes.viewModelsExt

private const val VISIBLE_THRESHOLD = 5

class JokesFragment : Fragment() {

    private lateinit var binding: JokesFragmentBinding
    private lateinit var recyclerAdapter: RecyclerAdapter

    private val viewModel by viewModelsExt {
        JokesViewModel()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View {
        binding = JokesFragmentBinding.inflate(inflater)
        val linearLayoutManager = LinearLayoutManager(context)
        binding.JokesList.layoutManager = linearLayoutManager
        recyclerAdapter = RecyclerAdapter(this)
        binding.JokesList.adapter = recyclerAdapter

        val dividerItemDecoration = DividerItemDecoration(binding.JokesList.context, linearLayoutManager.orientation)
        dividerItemDecoration.setDrawable(ResourcesCompat.getDrawable(resources, R.drawable.divider_item_shape, null)!!)
        binding.JokesList.addItemDecoration(dividerItemDecoration)

        binding.JokesList.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(unused: RecyclerView, dx: Int, dy: Int) {
                //вот это явно логика, но как её убрать отсюда? этот объект будет требовать постоянного проброса значений с linearLayoutManager
                if (linearLayoutManager.findLastVisibleItemPosition() + VISIBLE_THRESHOLD > linearLayoutManager.itemCount) {
                    viewModel.onScrolledToEnd()
                }
            }
        })

        viewModel.jokes.observe(viewLifecycleOwner) {
            recyclerAdapter.setData(it, JokesConstants.JOKES_VALUE)
        }

        viewModel.exceptionBundle.observe(viewLifecycleOwner) {
            (requireActivity() as MainActivity).changeFragmentByDirection(Shared.Direction.ERROR, it)
        }

        viewModel.onScrolledToEnd()

        return binding.JokesList
    }
}